articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+---
+layout: default
+title: Configuring QoS from an external file
+permalink: articles/qos_file.html
+abstract:
+  This article describes how to configure QoS from an external file.
+author: '[Ivan Santiago Paunovic](https://github.com/ivanpauno)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Overview
+
+`ROS 2` added the [Quality of Service (QoS)](qos.md) to publishers, subscriptions, clients and services.
+Up `ROS 2 Foxy`, the `QoS` can only be specified in code.
+To avoid recompiling the source code with patched `QoS`, node implementers have used different mechanisms to configure `QoS` profiles:
+- Command line arguments
+- ROS parameters
+- Combining system default QoS with vendor specific `QoS` profiles.
+
+Instead of having different ways of externally specifying `QoS` for the different endpoints, it would be good to have a standardized way to specific `QoS` profiles.
+
+## QoS file format
+
+DDS uses the concept of QoS profiles ([rti documentation](https://community.rti.com/examples/using-qos-profiles)), in which each profile has a name. When creating an entity (e.g. a `DataWriter`), the passed profile name will be looked up in the qos profiles file being used, and the QoS can be loaded in that way.
+
+ROS 2 can use a simpler approach, leveraging node name uniqueness <sup id="back_node_name_uniqueness">[1](#to_node_name_uniqueness)</sup>.
+
+<b id="to_node_name_uniqueness">1</b> Node name uniqueness is not being enforced up to Foxy, though it is supposed. [↩](#back_node_name_uniqueness)
+
+### Nodes
+
+In an acceptable `QoS` file format, it should be to distinguish `QoS` settings targeted to different nodes.
+Thus, a top level tag to identify the node should be available.
+Example alternatives:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
```

Comment:
> How will the namespacing work with remapping? Is the assumption that this must match a launched node?

Yes, the name and namespace should match the already remapped and expanded name of the node.

Parameter files work in the same way.

> Do we want to allow QoS profiles to be "remapped" to a node at different locations?

I think that it would be good in the file format to have a way to define a "profile", and then apply it to several entities.
I have to add that to the proposal.

Reply:
I have added the concept of "profile" in https://github.com/ros2/design/pull/296/commits/a0e635f3500e457bf56a36862651fe3bcf50b0a7.